apis/Google.LongRunning/Google.LongRunning/Operation.cs

```diff
@@ -105,17 +129,37 @@ public Operation(Operation rpcMessage, OperationsClient client)
         /// Retrieves the metadata associated with this operation, or <c>null</c> if there is no
         /// metadata in the underlying response message.
         /// </summary>
-        public TMetadata Metadata => RpcMessage.Metadata?.Unpack<TMetadata>();
+        /// <remarks>
+        /// Only the in-memory representation of the operation (this object) is consulted for its state.
+        /// See the <see cref="Operation"/> documentation for information about dealing with different metadata type versions.
+        /// </remarks>
+        /// <exception cref="InvalidOperationException">Metadata is present, but is not of the expected type.</exception>
+        public TMetadata Metadata => Unpack<TMetadata>(RpcMessage.Metadata, true);
 
         /// <summary>
-        /// Retrieves the result of the operation, throwing an exception if the operation failed or hasn't completed.
-        /// Unlike <see cref="Task{T}.Result"/>, this does not block.
+        /// Retrieves the metadata associated with this operation, or <c>null</c> if either there is no
+        /// metadata in the underlying response message, or it does not have the expected type.
         /// </summary>
         /// <remarks>
         /// Only the in-memory representation of the operation (this object) is consulted for its state.
+        /// See the <see cref="Operation"/> documentation for information about dealing with different metadata type versions.
+        /// </remarks>
+        /// <returns>The metadata of the operation if possible, or null otherwise.</returns>
+        public TMetadata GetMetadataOrNull() => Unpack<TMetadata>(RpcMessage.Metadata, false);
+
+        /// <summary>
+        /// Retrieves the result of the operation, throwing an exception if the operation failed, hasn't completed,
+        /// or has an unexpected result type. Unlike <see cref="Task{T}.Result"/>, this does not block.
+        /// If the operation has completed but the result is not present (for example due to being excluded by
+        /// a field mask) this returns null.
+        /// </summary>
+        /// <remarks>
+        /// Only the in-memory representation of the operation (this object) is consulted for its state.
+        /// See the <see cref="Operation"/> documentation for information about dealing with different response type versions.
         /// </remarks>
         /// <exception cref="OperationFailedException">The operation completed with an error.</exception>
-        /// <exception cref="InvalidOperationException">The operation has not completed yet.</exception>
+        /// <exception cref="InvalidOperationException">The operation has not completed yet, or the result is present but
```

Comment:
It feels a bit overkilly to me - especially as it's easy enough to check whether or not it *has* completed, if you need to. (You could even check it in a guard clause if you really wanted.)

InvalidOperationException really does feel like the most appropriate one...